plugins/__init__.py

```diff
@@ -82,57 +96,104 @@ def resource_attribute_reference(
     )
 
 
-@plugin
-def get_resource_attribute(
-    context: Context,
-    resource: "terraform::Resource",  # type: ignore
-    attribute_path: "any",  # type: ignore
-) -> "any":  # type: ignore
+def get_last_resource_parameter(
+    context: Context, resource: DynamicProxy, param_id: str, cache_dict: dict[str, dict]
+) -> dict:
     """
-    Get a resource attribute from the saved parameters (facts).
-    :param resource: The resource we which to get an attribute from.
-    :param attribute_path: The path, in the resource state dict, to the desired value.
+    Helper method to get the dict params from server or cache (if it is there).
+
+    :returns: The state dict found in parameters
+    :raises: UnknownStateException if the state dict is not found in parameters
     """
-    global resource_states
+    resource_id = inmanta.resources.to_id(resource)
 
-    attribute_reference = resource_attribute_reference(resource, attribute_path)
+    cached_parameter_dict = cache_dict.get(resource_id)
+    if cached_parameter_dict is not None:
+        return cached_parameter_dict
 
-    # Trying to get resource value from cache to avoid unnecessary api calls
-    cached_state = resource_states.get(attribute_reference.resource_id)
-    if cached_state is not None:
-        # Cache hit, we get the attribute value and return it
-        try:
-            return attribute_reference.extract_from_state(cached_state)
-        except ValueError as e:
-            raise PluginException(str(e))
+    environment = Config.get("config", "environment", None)
+    if environment is None:
+        raise PluginException(
+            "The environment for this model should be configured at this point"
+        )
 
     # Cache miss, we continue
     param_client = ParamClient(
-        environment=attribute_reference.environment,
+        environment=environment,
         client=context.get_client(),
         run_sync=lambda func: context.run_sync(func),  # type: ignore
-        param_id=TERRAFORM_RESOURCE_STATE_PARAMETER,
-        resource_id=attribute_reference.resource_id,
+        param_id=param_id,
+        resource_id=resource_id,
     )
 
-    resource_state_raw: Optional[str] = param_client.get()
-    if resource_state_raw is None:
+    resource_config_raw: Optional[str] = param_client.get()
+    if resource_config_raw is None:
         unknown_parameters.append(
             {
-                "resource": attribute_reference.resource_id,
-                "parameter": TERRAFORM_RESOURCE_STATE_PARAMETER,
+                "resource": resource_id,
+                "parameter": param_id,
                 "source": "fact",
             }
         )
-        return Unknown(source=resource)
+        raise UnknownStateException(Unknown(source=resource))
```

Comment:
How does this work the first time this model is compiled?

Reply:
The first time the model is compiled, and also every time the resource config is updated, the state dict attached to the config block will be unknown.  This doesn't prevent the resource from being deployed as this is not part of its attributes.  It might prevent other resources using those attributes to be deployed in this version as some of their attributes would then be unknown.

Reply:
won't this exception end the execution?

Reply:
No, this is a helper method, not a plugin, the plugins using it should catch this exception and handle the unknown as they want to.  The other approach I could have taken would be to return the unknown object, but it complicates the return type and requires some `isinstance` usage in the plugin using the helper.